text/0112-server-runtime.md

```diff
@@ -0,0 +1,331 @@
+---
+title: LWC server runtime
+status: DRAFTED
+created_at: 2020-01-06
+updated_at: 2020-01-06
+pr: (leave this empty until the PR is created)
+---
+
+# LWC server runtime
+
+## Summary
+
+The LWC engine has been designed from the beginning to run in an environment with access to the DOM APIs. To accommodate server-side rendering (SSR) requirements, we need a way to decouple the engine from the DOM APIs. The existing `@lwc/engine` will be replaced by 3 new packages: 
+- `@lwc/engine-core` exposes platform-agnostic APIs and will be used by the different runtime packages to share the common logic.
+- `@lwc/engine-dom` exposes LWC APIs available on the browser.
+- `@lwc/engine-server` exposes LWC APIs used for server-side rendering.
```

Comment:
Agreed, I wasn't happy with the naming. 

What do you think about `engine-dom-renderer` and `engine-string-renderer`? Even if the package names are more verbose I don't think it's a problem because the only thing developers have to remember if `lwc`.

Reply:
I also like the inverted version: `@lwc/client-engine`, `@lwc/server-engine`, `@lwc/engine`

Reply:
I don't think there will be a renaming impact here. The `@lwc/engine` module should never be referenced directly. Userland code should only reference `lwc`, the only place where we might need to change is on the resolver.

Reply:
I'm with @ekashida on this, this is NOT about client and server I'm afraid, this is about dom vs dom string. Imagine that in the future we decide to use the dom string output in a client for some reason.

Reply:
> [...] this is about dom vs dom string.

I fully agree with this statement. In this regard, I think it would less confusing if the module that is loaded in a DOM environment contains the term `dom` in its name. I would suggest reaming `@lwc/engine` to `@lwc/engine-dom` in @ekashida proposal.

Reply:
> I don't think there will be a renaming impact here. The @lwc/engine module should never be referenced directly.

I was talking about projects that have `@lwc/engine` as a dependency in their `package.json`. I'm assuming that DOM rendering will be the main use case for users so keeping that implicit in the package name seems like it would be less confusing.